src/pubsub/index.js

```diff
@@ -0,0 +1,663 @@
+    }
+
+    // Check the message signature if present
+    if (message.signature && !(await verifySignature(message))) {
+      throw errcode(new Error('Invalid message signature'), codes.ERR_INVALID_SIGNATURE)
+    }
+
+    for (const topic of message.topicIDs) {
+      const validatorFn = this.topicValidators.get(topic)
+      if (!validatorFn) {
+        continue
+      }
+      await validatorFn(topic, message)
+    }
+  }
+
+  /**
+   * Normalizes the message and signs it, if signing is enabled.
+   * Should be used by the routers to create the message to send.
+   * @private
+   * @param {Message} message
+   * @returns {Promise<Message>}
+   */
+  _buildMessage (message) {
+    const msg = utils.normalizeOutRpcMessage(message)
+    if (this.signMessages) {
+      return signMessage(this.peerId, msg)
+    } else {
+      return message
+    }
+  }
+
+  // API METHODS
+
+  /**
+   * Get a list of the peer-ids that are subscribed to one topic.
+   * @param {string} topic
+   * @returns {Array<string>}
+   */
+  getSubscribers (topic) {
+    if (!this.started) {
+      throw errcode(new Error('not started yet'), 'ERR_NOT_STARTED_YET')
+    }
+
+    if (!topic || typeof topic !== 'string') {
+      throw errcode(new Error('a string topic must be provided'), 'ERR_NOT_VALID_TOPIC')
+    }
+
+    const peersInTopic = this.topics.get(topic)
+    if (!peersInTopic) {
+      return []
+    }
+    return Array.from(peersInTopic)
+  }
+
+  /**
+   * Publishes messages to all subscribed peers
+   * @override
+   * @param {Array<string>|string} topics
+   * @param {Buffer} message
+   * @returns {Promise<void>}
+   */
+  async publish (topics, message) {
+    if (!this.started) {
+      throw new Error('Pubsub has not started')
+    }
+
+    topics = utils.ensureArray(topics)
+    this.log('publish', topics, message)
+
+    const from = this.peerId.toB58String()
+    let msgObject = {
+      receivedFrom: from,
+      from: from,
+      data: message,
+      seqno: utils.randomSeqno(),
+      topicIDs: topics
+    }
+
+    // Emit to self if I'm interested and emitSelf enabled
+    this.emitSelf && this._emitMessage(msgObject)
+
+    // ensure that any operations performed on the message will include the signature
+    msgObject = await this._buildMessage(msgObject)
+
+    // send to all the other peers
+    await this._publish(msgObject)
+  }
+
+  /**
+   * Overriding the implementation of publish should handle the appropriate algorithms for the publish/subscriber implementation.
+   * For example, a Floodsub implementation might simply publish each message to each topic for every peer
+   * @abstract
+   * @param {InMessage} message
+   * @returns {Promise<void>}
+   *
+   */
+  _publish (message) {
+    throw errcode(new Error('publish must be implemented by the subclass'), 'ERR_NOT_IMPLEMENTED')
+  }
+
+  /**
+   * Subscribes to a given topic.
+   * @abstract
+   * @param {string} topic
+   * @param {function} [handler]
```

Comment:
Is there somewhere where we can document what the input of this function will be? should be `(msg: InMessage) => void`

Reply:
good call! I can document it here